### PR TITLE
Refactor QgsMapLayer's encodedSource()/decodedSource()

### PR DESCRIPTION
**Pull request by wonder-sk**
Conversion between relative and absolute paths when reading/writing project files has been inconveniently done in subclasses of QgsMapLayer in encodedSource() and decodedSource() methods, with a lot of provider specific code.

This is a refactoring to finally fix this - there is a new pair of functions added to QgsProviderMetadata and QgsProviderRegistry:
- absoluteToRelativeUri()
- relativeToAbsoluteUri()

This new pair of functions delegates the actual conversion between absolute and relative path to provider code - where it belongs.

The existing code from encodedSource() and decodedSource() functions has been moved to providers. Unit tests were added to confirm the existing behavior.

There should be no change of behavior apart from removal of old compatibility code for WMS provider for projects from QGIS < 1.9 (10 years ago).

### CODE REVIEW

File: src/core/providers/ogr/qgsogrprovidermetadata.cpp [hunk 1]
```diff
@@ -316,6 +316,24 @@ QString QgsOgrProviderMetadata::encodeUri( const QVariantMap &parts ) const
   return uri;
 }
 
+QString QgsOgrProviderMetadata::absoluteToRelativeUri( const QString &uri, const QgsReadWriteContext &context ) const
+{
+  QString src = uri;
+  QStringList theURIParts = src.split( '|' );
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrprovidermetadata.cpp:322]: using decodeUri() and encodeUri() would avoid this method to know that | is the delimiter.
Same for relativeToAbsoluteUri()

File: src/core/pointcloud/qgspointcloudlayer.cpp [hunk 2]
```diff
@@ -484,32 +484,14 @@ QString QgsPointCloudLayer::encodedSource( const QString &source, const QgsReadW
...
-  if ( parts.contains( QStringLiteral( "path" ) ) )
-  {
-    parts.insert( QStringLiteral( "path" ), context.pathResolver().writePath( parts.value( QStringLiteral( "path" ) ).toString() ) );
-    return QgsProviderRegistry::instance()->encodeUri( providerType(), parts );
-  }
-  else
-  {
-    return source;
-  }
+  return QgsProviderRegistry::instance()->absoluteToRelativeUri( mProviderKey, source, context );
```
Comment (nyalldawson) [inline src/core/pointcloud/qgspointcloudlayer.cpp:487]: Could these implementations be moved to QgsMapLayer? They are shared by point cloud/ mesh/raster already.

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 3]
```diff
@@ -2454,6 +2454,172 @@ QString QgsGdalProviderMetadata::encodeUri( const QVariantMap &parts ) const
   return  QgsGdalProviderBase::encodeGdalUri( parts );
 }
 
+
+static bool _parseGpkgColons( const QString &src, QString &filename, QString &tablename )
+{
+  // GDAL accepts the following input format:  GPKG:filename:table
```
Comment (rouault) [inline src/core/providers/gdal/qgsgdalprovider.cpp:2460]: Seeing all this logic in QGIS makes me think that ideally, it should't belong to QGIS, but GDAL should have an API to ask drivers to do that job.

Comment (rouault) [reply src/core/providers/gdal/qgsgdalprovider.cpp:2460]: Tracked in https://github.com/OSGeo/gdal/issues/7261

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 4]
```diff
@@ -2454,6 +2454,172 @@ QString QgsGdalProviderMetadata::encodeUri( const QVariantMap &parts ) const
...
+
+
+QString QgsGdalProviderMetadata::absoluteToRelativeUri( const QString &uri, const QgsReadWriteContext &context ) const
+{
+  const QString src = uri;
+  if ( src.startsWith( QLatin1String( "NETCDF:" ) ) )
+  {
+    // NETCDF:filename:variable
+    // filename can be quoted with " as it can contain colons
+    const QRegularExpression netcdfEncodedRegExp( QRegularExpression::anchoredPattern( "NETCDF:(.+):([^:]+)" ) );
```
Comment (nyalldawson) [inline src/core/providers/gdal/qgsgdalprovider.cpp:2490]: May as well while we're at it:

```suggestion
    const thread_local QRegularExpression netcdfEncodedRegExp( QRegularExpression::anchoredPattern( "NETCDF:(.+):([^:]+)" ) );
```

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 5]
```diff
@@ -2454,6 +2454,172 @@ QString QgsGdalProviderMetadata::encodeUri( const QVariantMap &parts ) const
...
+    {
+      filename = context.pathResolver().writePath( filename );
+      return QStringLiteral( "GPKG:%1:%2" ).arg( filename, tablename );
+    }
+  }
+  else if ( src.startsWith( QLatin1String( "HDF4_SDS:" ) ) )
+  {
+    // HDF4_SDS:subdataset_type:file_name:subdataset_index
+    // filename can be quoted with " as it can contain colons
+    const QRegularExpression hdf4EncodedRegExp( QRegularExpression::anchoredPattern( "HDF4_SDS:([^:]+):(.+):([^:]+)" ) );
```
Comment (nyalldawson) [inline src/core/providers/gdal/qgsgdalprovider.cpp:2514]:
```suggestion
    const thread_local QRegularExpression hdf4EncodedRegExp( QRegularExpression::anchoredPattern( "HDF4_SDS:([^:]+):(.+):([^:]+)" ) );
```

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 6]
```diff
@@ -2454,6 +2454,172 @@ QString QgsGdalProviderMetadata::encodeUri( const QVariantMap &parts ) const
...
+      if ( filename.startsWith( '"' ) && filename.endsWith( '"' ) )
+        filename = filename.mid( 1, filename.length() - 2 );
+      return "HDF4_SDS:" + match.captured( 1 ) + ":\"" + context.pathResolver().writePath( filename ) + "\":" + match.captured( 3 );
+    }
+  }
+  else if ( src.startsWith( QLatin1String( "HDF5:" ) ) )
+  {
+    // HDF5:file_name:subdataset
+    // filename can be quoted with " as it can contain colons
+    const QRegularExpression hdf5EncodedRegExp( QRegularExpression::anchoredPattern( "HDF5:(.+):([^:]+)" ) );
```
Comment (nyalldawson) [inline src/core/providers/gdal/qgsgdalprovider.cpp:2528]:
```suggestion
    const thread_local QRegularExpression hdf5EncodedRegExp( QRegularExpression::anchoredPattern( "HDF5:(.+):([^:]+)" ) );
```

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 7]
```diff
@@ -2454,6 +2454,172 @@ QString QgsGdalProviderMetadata::encodeUri( const QVariantMap &parts ) const
...
+      if ( filename.startsWith( '"' ) && filename.endsWith( '"' ) )
+        filename = filename.mid( 1, filename.length() - 2 );
+      return "HDF5:\"" + context.pathResolver().writePath( filename ) + "\":" + match.captured( 2 );
+    }
+  }
+  else if ( src.contains( QRegularExpression( "^(NITF_IM|RADARSAT_2_CALIB):" ) ) )
+  {
+    // NITF_IM:0:filename
+    // RADARSAT_2_CALIB:?:filename
+    const QRegularExpression nitfRadarsatEncodedRegExp( QRegularExpression::anchoredPattern( "([^:]+):([^:]+):(.+)" ) );
```
Comment (nyalldawson) [inline src/core/providers/gdal/qgsgdalprovider.cpp:2542]:
```suggestion
    const thread_local QRegularExpression nitfRadarsatEncodedRegExp( QRegularExpression::anchoredPattern( "([^:]+):([^:]+):(.+)" ) );
```

**Comment by nyalldawson** [comment]
Great move! I suspect there's lots of cleanup/modernisation we can do to the individual implementations, but I'm fine with deferring that for now.

File: src/core/pointcloud/qgspointcloudlayer.cpp (hunk 2 above)
Comment (wonder-sk) [reply src/core/pointcloud/qgspointcloudlayer.cpp:487]: I thought about doing that, but in the end decided to skip that for now:
- base map layer class does not do any absolute/relative path conversion - changing that could in theory cause issue(s) in custom plugin layers (?)
- vector layer has an extra odd bit of code for memory layers
- vector tile layer needs forcing provider key to be `vectortile` as earlier versions of QGIS did saved empty provider key

File: src/core/providers/gdal/qgsgdalprovider.cpp (hunk 3 above)
Comment (wonder-sk) [reply src/core/providers/gdal/qgsgdalprovider.cpp:2460]: That would be quite nice if GDAL had such API... as you can see in QGIS code we're only handling few (most common?) cases

File: src/core/providers/ogr/qgsogrprovidermetadata.cpp (hunk 1 above)
Comment (wonder-sk) [reply src/core/providers/ogr/qgsogrprovidermetadata.cpp:322]: Indeed... there are a couple more improvements I can immediately see (e.g. handling of raster/vector virtual layer providers), but I wanted to keep this PR with least amount of changes, only move the code around, and leave logic improvements for later

File: src/core/pointcloud/qgspointcloudlayer.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/pointcloud/qgspointcloudlayer.cpp:487]: I'm fine with that!

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**